language/vm/src/binary_format.rs

```diff
@@ -0,0 +1,362 @@
+
+    // Below should be enough information to materialize a
+    // main<Cup<T>, W>(arg1, ..., argn) {}
+    // with args of any type (including generic types) that the VM (verification
+    // and runtime) can use.
+    // There is a verification step here that is *purely* runtime.
+    // The client is passing the type arguments and you can imagine restrictions
+    // given our resource semantic.
+
+    // function arguments signature, main is void on return
+    pub arg_types: SignatureIndex,
+    // type parameter for the script
+    pub type_params: Vec<Kind>,
+    // code
+    pub main: CodeUnit,
+}
+
+// A CompiledModule is what it used to be but you will notice the removal of `FieldDefinition`s.
+// Fields are now serialized with the `StructDefinition` which changes verification around fields.
+// (see `StructDefinition` for details)
+pub struct CompiledModule {
+    pub header: CompiledHeader,
+
+    pub struct_defs: Vec<StructDefinition>,
+    pub function_defs: Vec<FunctionDefinition>,
+}
+
+//
+// Basics...
+//
+
+pub enum Kind {
+    All,
+    Resource,
+    Unrestricted,
+}
+
+// Any expressible type in Move
+pub enum Type {
+    Bool,
+    U8,
+    U64,
+    U128,
+    Address,
+    Vector(Box<Type>),
+    Struct(StructHandleIndex),
+    // a complete or partial instantiation of a generic type
+    StructSpec(StructSpecIndex),
+    Reference(Box<Type>),
+    MutableReference(Box<Type>),
+    // a type argument provided by the definition (via instantiation)
+    TypeParameter(TypeParameterIndex),
+}
+
+//
+// Handles
+//
+
+
+pub struct ModuleHandle {
+    pub address: AddressPoolIndex,
+    pub name: IdentifierIndex,
+}
+
+
+pub struct StructHandle {
+    pub module: ModuleHandleIndex,
+    pub name: IdentifierIndex,
+    pub is_nominal_resource: bool,
+    pub type_params: Vec<Kind>,
+}
+
+// FunctionHandles inlines the signature as indexes into the sig table
+pub struct FunctionHandle {
+    pub module: ModuleHandleIndex,
+    pub name: IdentifierIndex,
+    pub arg_types: SignatureIndex,
+    pub return_types: SignatureIndex,
+    pub type_params: Vec<Kind>,
+}
+
+// A field access info (owner type and offset)
+pub struct FieldHandle {
+    owner: StructDefinitionIndex,
+    field: FieldOffset,
+}
+
+//
+// Specifications
+//
+// Specifications point to a generic handle and its instantiation.
+// The instantiation can be partial.
+// So for types a StructSpec is either S<T, W>, S<u8, bool>, S<T, u8>
+
+// A complete or partial instantiation of a generic struct
+pub struct StructSpec {
+    pub handle: StructHandleIndex,
+    pub instantiation: SignatureIndex,
+}
+
+// A complete or partial instantiation of a function
+pub struct FunctionSpec {
+    pub handle: FunctionHandleIndex,
+    pub instantiation: SignatureIndex,
+}
+
+// A complete or partial instantiation of a field (or the type of it).
+// A FieldSpec points to a generic FieldHandle and the owner type instantiation.
+pub struct FieldSpec {
```

Comment:
a `FieldSpec` now points to a `FieldHandle` + instantiation. 
The `FieldHandle` has been changed to point to a `StructDefinition` so we get back the guarantee at binary format that the field can only be accessed within a module.
@tnowacki  can you comment about the compiler generating this table? 
Compared with the previous iteration this basically splits the type instantiation by having the generic type in the `FieldHandle` and the instantiation in the `FieldSpec`.
That is consistent with the other specification but less with the usage of a "type"

Reply:
I am not sure I follow or understand the question/comment well so let me try to explain better and I definitely realize this needs a more explicit comment.
The `instantiation` in the `FieldSpec` is that of the type in the `FieldHandle`, so for `S<u8, bool>.f` where `f` is a field of any type the `instantiation` in  the `FieldSpec` would be `[u8, boo]`, that is, the instantiation for `S`.
I am not sure if that is what you wanted it to be (and it is) or the other way around :)

Reply:
> S<u8, bool>.f where f is a field of any type the instantiation in the FieldSpec would be [u8, boo], that is, the instantiation for S.

This is what I expected. But the comments confused me